File: ts/src/ranking.ts
// Importance ranking module: PageRank for "the N most important symbols" (full TypeScript rewrite,
// P1; ported from Python codesextant/ranking.py).
//
// Design, taken from the PoC and aider's repomap: defined symbols are graph nodes and reference edges
// (who uses whom) are the links; the more symbols that are themselves important reference a symbol,
// the more important that symbol is (PageRank's recursive definition). Plain power iteration, with no
// networkx or scipy dependency, which keeps the engine light.
//
// Porting rules, checked against the frozen Python version's golden-test ground truth
// (test/fixtures/expected_ranking.json):
//   - IEEE 754 doubles behave identically in both languages; as long as the *order* of operations
//     lines up statement for statement, the scores match to a very small tolerance.
//   - The symbol_id string format (path::scope::name::line) is identical to the Python f-string, so
//     the keys of scores can be compared across languages.
//   - dict/Counter → Map, list → array. Iteration order (a Map keeps insertion order, an array keeps
//     push order) matches Python, which keeps floating-point accumulation order identical.
//   - Environment parsing follows Python's float()/int() strictly: an empty or non-numeric string
//     falls back to the default, and int refuses a string with a decimal point.
//
// Architecture note (temporary for P1): importing storage.normPath reuses the single source of truth
//   for "match Python's normcase(abspath())": a reference edge's def_path/src_path and a symbol's
//   path have to go through the same normalization to land on the same node. The cost is that loading
//   this module pulls in storage, and with it better-sqlite3; the ranking code itself makes no sqlite
//   calls and remains pure algorithm. Once namegraph.ts needs the same normalization, extract it into
//   a shared paths.ts.
//
// Single responsibility: take a list of symbols and a list of reference edges, return the symbols
// carrying a rank score, sorted high to low. It touches neither SQLite nor ts-morph. All state is
// local to the functions, so it is re-entrant and pollutes nothing global.
import { normPath } from "./storage.js";

// Weights for reference edges by confidence. A link confirmed by jedi or ts-morph is more
// trustworthy than a name match, so it weighs more.
const CONFIDENCE_WEIGHT: Record<string, number> = { high: 1.0, low: 0.25 };

/** A symbol going into ranking. Looser than storage.SymbolRow: path, end_line and scope may all be
 *  absent, matching the dict.get tolerance on the Python side.
 *  The index signature is what lets rank_symbols return {...s, rank} with every other field intact. */
export interface RankSymbol {
  name: string;
  line: number;
  path?: string | null;
  scope?: string | null;
  end_line?: number | null;
  kind?: string | null;
  [extra: string]: unknown;
}

/** A reference edge going into ranking. Every field is optional, matching Python's e.get(...)
 *  tolerance; def_path and def_line are null when the definition was never resolved. */
export interface RankRef {
  def_path?: string | null;
  def_line?: number | null;
  src_path?: string | null;
  src_line?: number | null;
  confidence?: string;
}

// ── env helpers (following Python's float()/int() semantics exactly: an invalid string falls back to
//    the default) ──

/** Matches Python's float(os.environ.get(name, "")): an empty or non-numeric string falls back to the
 *  default, which is where Python raises ValueError. */
function envFloat(name: string, def: number): number {
  const raw = process.env[name];
  if (raw === undefined || raw.trim() === "") return def; // float("") and float("  ") both raise
  const v = Number(raw);
  return Number.isNaN(v) ? def : v; // "1.5x" → NaN → default (Python raises here)
}

/** Matches Python's int(os.environ.get(name, "")): only integer strings are accepted, so "5.5", "",
 *  "abc" and "0x1f" all fall back to the default. */
function envInt(name: string, def: number): number {
  const raw = process.env[name];
  if (raw === undefined) return def;
  const t = raw.trim();
  if (!/^[+-]?\d+$/.test(t)) return def; // Python's int() raises on decimal points and non-numeric strings
  return parseInt(t, 10);
}

/** A public symbol that follows a naming convention: it contains an underscore separator or mixes
 *  case, i.e. snake_case, camelCase or PascalCase. */
export function wellNamed(name: string): boolean {
  if (name.startsWith("_")) return false;
  return name.includes("_") || (name !== name.toLowerCase() && name !== name.toUpperCase());
}

/** queue 5 (symbol quality factor on edge weights, an aider heuristic): bring architectural public
 *  APIs forward and push low-signal or ubiquitous symbols back.
 *  A well-named public symbol (length >= the threshold) gets ×WELLNAMED; a private one starting with
 *  _ gets ×PRIVATE; one defined in more than N files (a ubiquitous name) gets ×COMMON. */
export function symbolQualityMult(name: string, definesCount: number): number {
  let mult = 1.0;
  if (name.startsWith("_")) {
    mult *= envFloat("CODESEXTANT_RANK_PRIVATE_MULT", 0.1);
  } else if (name.length >= envInt("CODESEXTANT_RANK_WELLNAMED_MINLEN", 8) && wellNamed(name)) {
    mult *= envFloat("CODESEXTANT_RANK_WELLNAMED_MULT", 10.0);
  }
  if (definesCount > envInt("CODESEXTANT_RANK_COMMON_THRESHOLD", 5)) {
    mult *= envFloat("CODESEXTANT_RANK_COMMON_MULT", 0.1);
  }
  return mult;
}

/** A symbol's unique id: path::scope::name::line. Including line keeps same-named symbols in one file
 *  from overwriting each other.
 *  ⚠ Identical to the Python f-string; the golden test's scores keys depend on it. */
export function symbolId(sym: RankSymbol): string {
  return `${sym.path}::${sym.scope ?? ""}::${sym.name}::${sym.line}`;
}

/** Path normalization (matches Python's os.path.normcase(os.path.abspath()) if path else ""). */
function norm(path: string | null | undefined): string {
  return path ? normPath(path) : "";
}

/** Pull the line out of a symbol_id (format path::scope::name::line). Returns a very large value when
 *  it cannot be parsed, matching Python's 1<<30. */
function lineOf(sid: string): number {
  const i = sid.lastIndexOf("::");
  if (i === -1) return 1 << 30;
  const v = Number(sid.slice(i + 2));
  return Number.isNaN(v) ? 1 << 30 : v;
}

/** queue 4 (query-aware PageRank): turn the focus set the caller passes in into a personalization
 *  vector.
 *  ⛔ The boost comes from the caller explicitly stating "I am working on X" (focusSymbols/
 *  focusFiles), never from listening to a conversation or calling an LLM. The zero-cloud rule holds.
 *  Symbols hit by the focus get boost added to their teleport weight. With no focus this returns
 *  null, which falls back to uniform teleport, the original static behaviour. */
export function buildPersonalization(
  symbols: RankSymbol[],
  focusSymbols?: string[] | null,
  focusFiles?: string[] | null,
): Map<string, number> | null {
  const fs = new Set(focusSymbols ?? []);
  const ff = new Set((focusFiles ?? []).map((f) => norm(f)));
  if (fs.size === 0 && ff.size === 0) return null;
  const boost = envFloat("CODESEXTANT_PAGERANK_FOCUS_BOOST", 10.0);
  const p = new Map<string, number>();
  for (const s of symbols) {
    let w = 1.0;
    if (fs.has(s.name)) w += boost;
    if (ff.has(norm(s.path))) w += boost;
    p.set(symbolId(s), w);
  }
  return p;
}

/** Run PageRank over the symbol graph and return Map{symbol_id → score}.
 *
 *  Edge direction: src (the representative symbol of the file the reference sits in) → def (the
 *  referenced symbol's definition). PageRank flows score from the referencing side to the referenced
 *  side, so a symbol referenced by many important symbols scores high. When src maps to no node (a
 *  module-top-level call, for instance), it is counted as evenly-shared external inflow.
 *  queue 5 multiplies the edge weight by the referenced symbol's quality factor. queue 4 feeds
 *  personalization into teleport, which is what makes it query-aware; null falls back to uniform.
 *  An empty symbols list returns an empty Map. Every intermediate is local, so calls cannot affect
 *  each other. */
export function computePagerank(
  symbols: RankSymbol[],
  refs: RankRef[],
  opts: {
    damping?: number;
    maxIter?: number;
    tol?: number;
    personalization?: Map<string, number> | null;
  } = {},
): Map<string, number> {
  const { damping = 0.85, maxIter = 100, tol = 1.0e-6, personalization = null } = opts;
  if (symbols.length === 0) return new Map();

  const nodeIds = symbols.map((s) => symbolId(s));
  const n = nodeIds.length;
  const idx = new Map<string, number>();
  nodeIds.forEach((sid, i) => idx.set(sid, i)); // on a duplicate id the later write wins (matching the Python dict comprehension)
  const nameOf = new Map<string, string>();
  nodeIds.forEach((sid, i) => nameOf.set(sid, symbols[i]!.name));

  // queue 5: name → the number of distinct (name, normPath) definitions it has. Too common means low
  // signal, and the edge weight is discounted accordingly.
  const seenNp = new Set<string>();
  const defines = new Map<string, number>();
  for (const s of symbols) {
    const k = `${s.name} ${norm(s.path)}`;
    if (!seenNp.has(k)) {
      seenNp.add(k);
      defines.set(s.name, (defines.get(s.name) ?? 0) + 1);
    }
  }

  // (absolute file path, definition line) → symbol_id, so a reference edge's def_path/def_line can
  // find its node; absolute file path → the symbol_id of that file's first symbol, used as the
  // representative node for the referencing side.
  const byPos = new Map<string, string>();
  const fileRep = new Map<string, string>();
  for (const s of symbols) {
    const p = s.path;
    if (p === null || p === undefined) continue; // matches Python's `if p is None: continue` (an empty string is not skipped)
    const sid = symbolId(s);
    const np = norm(p);
    byPos.set(`${np} ${s.line}`, sid);
    const cur = fileRep.get(np);
    if (cur === undefined || s.line < lineOf(cur)) fileRep.set(np, sid);
  }

  // path → sorted [(line, end_line, sid)], used to map a src_line to the caller symbol containing it.
  // Red-team L1-HIGH fix (made on the Python side): for an edge inside one file, use the caller that
  // actually contains src_line as the node instead of collapsing to the file's first symbol
  // (fileRep). Otherwise, whenever the callee happens to be the file's first symbol, the edge
  // becomes i==j and is skipped, leaving same-file reference structure invisible to PageRank.
  const byBody = new Map<string, Array<[number, number, string]>>();
  for (const s of symbols) {
    const p = s.path;
    if (p === null || p === undefined) continue;
    const np = norm(p);
    const el = Number(s.end_line ?? s.line) || s.line; // matches int(s.get("end_line", line) or line)
    let arr = byBody.get(np);
    if (arr === undefined) {
      arr = [];
      byBody.set(np, arr);
    }
    arr.push([s.line, el, symbolId(s)]);
  }
  for (const arr of byBody.values()) {
    // Matches the Python tuple sort: (line, end_line, sid) lexicographic. sid compares by code unit,
    // which orders identically to code point for anything in the BMP.
    arr.sort((a, b) => a[0] - b[0] || a[1] - b[1] || (a[2] < b[2] ? -1 : a[2] > b[2] ? 1 : 0));
  }

  // Map a src_line to the innermost symbol containing it and use that as the source node; with no
  // src_line, or when nothing matches, fall back to fileRep.
  const srcNode = (
    srcPath: string | null | undefined,
    srcLine: number | null | undefined,
  ): string | undefined => {
    const np = srcPath ? norm(srcPath) : "";
    if (srcLine && byBody.has(np)) {
      let best: string | undefined = undefined;
      for (const [ln, el, sid] of byBody.get(np)!) {
        if (ln > srcLine) break;
        if (ln <= srcLine && srcLine <= el) best = sid; // keep updating → ends on the innermost one (largest line)
      }
      if (best !== undefined) return best;
    }
    return fileRep.get(np);
  };

  // Build the adjacency outTargets[i] = [(j, weight), ...], plus external inflow in externalInflow[j].
  const outTargets: Array<Array<[number, number]>> = Array.from({ length: n }, () => []);
  const externalInflow = new Map<number, number>();

  for (const e of refs) {
    const dp = e.def_path ?? null;
    const dl = e.def_line ?? null;
    if (dp === null || dl === null) continue;
    const target = byPos.get(`${norm(dp)} ${dl}`);
    if (target === undefined) continue;
    const j = idx.get(target)!;
    let w = CONFIDENCE_WEIGHT[e.confidence ?? "low"] ?? 0.25;
    // queue 5: multiply in the referenced symbol's quality factor.
    const tname = nameOf.get(target) ?? "";
    w *= symbolQualityMult(tname, defines.get(tname) ?? 1);

    const srcRep = srcNode(e.src_path, e.src_line);
    if (srcRep === undefined) {
      externalInflow.set(j, (externalInflow.get(j) ?? 0.0) + w);
      continue;
    }
    const i = idx.get(srcRep)!;
    if (i === j) continue;
    outTargets[i]!.push([j, w]);
  }

  const nRefs = Math.max(1, refs.length);
  // queue 4: the personalization teleport vector P (the focus preference); without one, a uniform
  // 1/n, which is the original static behaviour and stays backward compatible.
  let P: number[];
  if (personalization !== null && personalization.size > 0) {
    let totP = 0;
    for (const sid of nodeIds) totP += personalization.get(sid) ?? 1.0;
    if (totP === 0) totP = 1.0;
    P = nodeIds.map((sid) => (personalization.get(sid) ?? 1.0) / totP);
  } else {
    P = new Array<number>(n).fill(1.0 / n);
  }
  let score = [...P];

  for (let iter = 0; iter < maxIter; iter++) {
    const newScore = P.map((pj) => (1.0 - damping) * pj);
    let danglingSum = 0.0;
    for (let i = 0; i < n; i++) {
      const edges = outTargets[i]!;
      if (edges.length === 0) {
        danglingSum += score[i]!;
        continue;
      }
      let totalW = 0;
      for (const [, w] of edges) totalW += w;
      if (totalW <= 0) {
        danglingSum += score[i]!;
        continue;
      }
      for (const [j, w] of edges) newScore[j]! += damping * score[i]! * (w / totalW);
    }
    if (danglingSum) {
      // Dangling mass flows back along the personalization distribution (aider uses dangling=P).
      for (let j = 0; j < n; j++) newScore[j]! += damping * danglingSum * P[j]!;
    }
    for (const [j, infl] of externalInflow) {
      newScore[j]! += (damping * infl) / nRefs;
    }

    let delta = 0;
    for (let i = 0; i < n; i++) delta += Math.abs(newScore[i]! - score[i]!);
    score = newScore;
    if (delta < tol) break;
  }

  const out = new Map<string, number>();
  for (let i = 0; i < n; i++) out.set(nodeIds[i]!, score[i]!);
  return out;
}

/** Rank symbols by importance and return them carrying a "rank" score, sorted high to low (the
 *  original fields plus rank).
 *  Passing topN returns only the first N. focusSymbols/focusFiles (queue 4, query-aware) are the
 *  symbols and files the caller explicitly says are being edited or asked about, which biases the
 *  ranking toward them; omit them for the original static structural-centrality ranking. */
export function rankSymbols<T extends RankSymbol>(
  symbols: T[],
  refs: RankRef[],
  opts: {
    topN?: number | null;
    damping?: number;
    focusSymbols?: string[] | null;
    focusFiles?: string[] | null;
  } = {},
): Array<T & { rank: number }> {
  const { topN = null, damping = 0.85, focusSymbols = null, focusFiles = null } = opts;
  const personalization = buildPersonalization(symbols, focusSymbols, focusFiles);
  const scores = computePagerank(symbols, refs, { damping, personalization });
  const ranked = symbols.map((s) => ({ ...s, rank: scores.get(symbolId(s)) ?? 0.0 }));
  // Stable descending sort: equal ranks keep their original relative order, matching the stability of
  // Python's sort(reverse=True).
  ranked.sort((a, b) => b.rank - a.rank);
  return topN !== null ? ranked.slice(0, topN) : ranked;
}

File: codesextant/ranking.py
"""Importance ranking module: PageRank for "the N most important symbols".

Design origin (borrowed from the PoC / aider repomap approach):
  - Treat each "defined symbol" as a graph node, and each reference edge
    (who uses whom) as a link.
  - A symbol referenced by more symbols that are themselves important is
    more important (PageRank's recursive definition).
  - aider's repomap uses the same graph-rank idea to pick which symbols
    are worth showing the LLM within a token budget.

The implementation deliberately uses plain Python power iteration instead
of pulling in a networkx/scipy dependency (keeps the engine lightweight
and easy to bundle into the daemon; power iteration at this scale is fast
enough even with tens of thousands of symbols).

Single responsibility: take a symbol list + a reference-edge list, emit
symbols sorted high to low by rank score. Does not touch SQLite, does not
touch jedi. All state is local to the functions, which makes them reentrant and
free of global pollution.
"""
from __future__ import annotations

import os
from bisect import bisect_right
from collections import Counter
from heapq import nlargest

# Weight of high-confidence reference edges (a jedi-confirmed target is more trustworthy
# than a name match, so it gets a higher weight)
_CONFIDENCE_WEIGHT = {"high": 1.0, "low": 0.25}


def _env_float(name: str, default: float) -> float:
    try:
        return float(os.environ.get(name, ""))
    except ValueError:
        return default


def _env_int(name: str, default: int) -> int:
    try:
        return int(os.environ.get(name, ""))
    except ValueError:
        return default


def _well_named(name: str) -> bool:
    """Well-formed public symbol (underscore-separated or mixed case = snake/camel/Pascal)."""
    if name.startswith("_"):
        return False
    return ("_" in name) or (name != name.lower() and name != name.upper())


def _symbol_quality_mult(name: str, defines_count: int) -> float:
    """queue 5 (edge-weight symbol-quality factor, an aider-inspired heuristic): surfaces
    architecturally significant public APIs and downweights low-signal/generic symbols.

    Well-named public symbols (len>=threshold) get ×WELLNAMED; private, underscore-prefixed
    symbols get ×PRIVATE; symbols redefined in >N files (overly generic names like
    utils/handle/run) get ×COMMON. All configurable (L0 hard rule #6).
    """
    mult = 1.0
    if name.startswith("_"):
        mult *= _env_float("CODESEXTANT_RANK_PRIVATE_MULT", 0.1)
    elif len(name) >= _env_int("CODESEXTANT_RANK_WELLNAMED_MINLEN", 8) and _well_named(name):
        mult *= _env_float("CODESEXTANT_RANK_WELLNAMED_MULT", 10.0)
    if defines_count > _env_int("CODESEXTANT_RANK_COMMON_THRESHOLD", 5):
        mult *= _env_float("CODESEXTANT_RANK_COMMON_MULT", 0.1)
    return mult


def _build_personalization(symbols: list[dict], focus_symbols=None,
                           focus_files=None) -> dict | None:
    """queue 4 (query-aware PageRank): turns a caller-supplied focus set into a
    personalization vector.

    ⛔ The boost NEVER comes from listening to a conversation or calling an LLM. It only
    comes from the caller explicitly saying "I'm working on X" (focus_symbols/focus_files).
    This holds the zero-cloud / no-LLM hard rule: aider listens to chat because it is a chat
    frontend, CodeSextant is a tool that gets called, not one that eavesdrops.
    Symbols matching focus get their teleport weight boosted by `boost`x. No focus returns
    None (falls back to uniform teleport = the original static behavior).
    """
    fs = set(focus_symbols or [])
    ff = {_norm(f) for f in (focus_files or [])}
    if not fs and not ff:
        return None
    boost = _env_float("CODESEXTANT_PAGERANK_FOCUS_BOOST", 10.0)
    p: dict[str, float] = {}
    for s in symbols:
        w = 1.0
        if s.get("name") in fs:
            w += boost
        if _norm(s.get("path")) in ff:
            w += boost
        p[_symbol_id(s)] = w
    return p


def _symbol_id(sym: dict) -> str:
    """The symbol's unique identifier: path::scope::name::line.
    Including the line in the id avoids same-file same-name symbols (e.g. multiple
    getters/setters sharing a name) overwriting each other.
    """
    return f"{sym['path']}::{sym.get('scope', '')}::{sym['name']}::{sym['line']}"


def _norm(path: str | None) -> str:
    return os.path.normcase(os.path.abspath(path)) if path else ""


def _compute_pagerank_scores(symbols: list[dict], refs: list[dict],
                             *, damping: float = 0.85, max_iter: int = 100,
                             tol: float = 1.0e-6,
                             personalization: dict[str, float] | None = None
                             ) -> list[float]:
    """Run PageRank on the symbol graph, return a score list in the same order as symbols.

    Edge direction: src (the representative symbol of the file containing the reference) ->
    def (the referenced symbol's definition). PageRank flows score from the referencing end
    to the referenced end, so a symbol referenced by many important symbols scores high.
    When src can't be mapped to a symbol node (e.g. a module-level top-level call), it is
    counted as "external inflow" distributed evenly.

    queue 5: edge weight is further multiplied by the referenced symbol's quality factor
    (well-named public ×10 / private ×0.1 / overly generic ×0.1).
    queue 4: personalization ({symbol_id: preference weight}) feeds the teleport vector for
    query-aware ranking; None falls back to uniform teleport (the original static behavior,
    backward compatible).

    Returns [] for empty symbols. Uses list index as the internal node id; the public
    compute_pagerank only converts to string ids at the very end, while rank_symbols
    consumes the list directly, which avoids building two giant string dicts for a graph
    with 570K nodes.
    """
    if not symbols:
        return []

    n = len(symbols)

    # namegraph can already fold repeated occurrences on the same line via multiplicity;
    # older db edges lacking this field are treated as 1.
    # Aggregating before building the graph preserves the original weights while letting
    # everything downstream work with unique edges only.
    collapsed_refs: dict[tuple, int] = {}
    for e in refs:
        key = (e.get("src_path"), e.get("src_line"), e.get("def_path"),
               e.get("def_line"), e.get("confidence", "low"))
        try:
            multiplicity = int(e.get("multiplicity", 1) or 1)
        except (TypeError, ValueError):
            multiplicity = 1
        if multiplicity < 1:
            multiplicity = 1
        collapsed_refs[key] = collapsed_refs.get(key, 0) + multiplicity

    # Path normalization is not cheap on Windows; 570K symbols are typically spread across
    # tens of thousands of files, so caching by the original string avoids the same path
    # being recomputed by os.path.abspath dozens of times across the defines/by_pos/by_body
    # layers.
    norm_cache: dict[object, str] = {}

    def _norm_cached(path) -> str:
        if not path:
            return ""
        try:
            return norm_cache[path]
        except KeyError:
            value = _norm(path)
            norm_cache[path] = value
            return value

    target_positions = {
        (_norm_cached(dp), dl)
        for (_sp, _sl, dp, dl, _confidence) in collapsed_refs
        if dp is not None and dl is not None
    }
    source_paths = {
        _norm_cached(sp)
        for (sp, _sl, _dp, _dl, _confidence) in collapsed_refs
        if sp
    }

    # Build the position tables only for the target/source of actual reference edges. The
    # old version built by_pos/by_body for all 570K symbols, paying the full-graph dict/list
    # cost even when the graph only had a few thousand edges.
    by_pos: dict[tuple, int] = {}
    target_name_of: dict[int, str] = {}
    file_rep: dict[str, int] = {}
    file_rep_line: dict[str, int] = {}
    by_body: dict[str, list] = {}
    for pos, s in enumerate(symbols):
        p = s.get("path")
        if p is None:
            continue
        np = _norm_cached(p)
        line = s["line"]
        if (np, line) in target_positions:
            by_pos[(np, line)] = pos
            target_name_of[pos] = s["name"]
        if np in source_paths:
            if np not in file_rep_line or line < file_rep_line[np]:
                file_rep[np] = pos
                file_rep_line[np] = line
            by_body.setdefault(np, []).append(
                (line, int(s.get("end_line", line) or line), pos))
    for lst in by_body.values():
        lst.sort()
    body_starts = {path: [row[0] for row in rows] for path, rows in by_body.items()}
    src_node_cache: dict[tuple[str, int], int | None] = {}

    # queue 5: only compute, for names that actually become edge targets, how many distinct
    # files define them; the other 570K nodes never use the quality factor, so there's no
    # need to build a full name->file set.
    target_names = set(target_name_of.values())
    _seen_np: set = set()
    defines: Counter[str] = Counter()
    if target_names:
        for s in symbols:
            name = s["name"]
            if name not in target_names:
                continue
            k = (name, _norm_cached(s.get("path")))
            if k not in _seen_np:
                _seen_np.add(k)
                defines[name] += 1

    def _src_node(src_path, src_line):
        """Maps src_line to "the innermost symbol that contains it" as the source node; no
        src_line / not found -> falls back to file_rep (backward compatible: db
        high-confidence edges carry src_line = a more precise caller; src_line=0 falls back
        to file_rep)."""
        np = _norm_cached(src_path) if src_path else ""
        cache_key = (np, int(src_line or 0))
        if cache_key in src_node_cache:
            return src_node_cache[cache_key]
        if src_line and np in by_body:
            rows = by_body[np]
            pos = bisect_right(body_starts[np], src_line) - 1
            while pos >= 0:
                ln, el, node_index = rows[pos]
                if ln <= src_line <= el:
                    src_node_cache[cache_key] = node_index
                    return node_index
                pos -= 1
        result = file_rep.get(np)
        src_node_cache[cache_key] = result
        return result

    # Build the sparse weighted adjacency out_targets[i] = {j: summed_weight}; only sources
    # with an edge occupy a dict entry.
    # namegraph keeps every occurrence to express reference counts, so the same
    # caller->target pair can end up with tens of thousands of duplicate edges. PageRank
    # only needs the total weight: aggregating first preserves the exact same math while
    # avoiding walking every occurrence on every iteration (a large TS repo once made
    # /get_map exceed the client's 30s timeout because of this).
    out_targets: dict[int, dict[int, float]] = {}
    external_inflow: dict[int, float] = {}
    quality_cache: dict[tuple[str, int], float] = {}

    for (src_path, src_line, dp, dl, confidence), multiplicity in collapsed_refs.items():
        if dp is None or dl is None:
            continue
        j = by_pos.get((_norm_cached(dp), dl))
        if j is None:
            continue
        w = _CONFIDENCE_WEIGHT.get(confidence, 0.25) * multiplicity
        # queue 5: multiply in the referenced symbol's quality factor (well-named public
        # ×10 / private ×0.1 / overly generic ×0.1)
        tname = target_name_of.get(j, "")
        quality_key = (tname, defines.get(tname, 1))
        try:
            quality = quality_cache[quality_key]
        except KeyError:
            quality = _symbol_quality_mult(*quality_key)
            quality_cache[quality_key] = quality
        w *= quality

        i = _src_node(src_path, src_line)
        if i is None:
            external_inflow[j] = external_inflow.get(j, 0.0) + w
            continue
        if i == j:
            continue
        edges = out_targets.setdefault(i, {})
        edges[j] = edges.get(j, 0.0) + w

    n_refs = max(1, sum(collapsed_refs.values()))
    # queue 4: personalization teleport vector P (focus preference); otherwise uniform 1/n
    # (original static behavior, backward compatible)
    if personalization:
        raw_p = [personalization.get(_symbol_id(s), 1.0) for s in symbols]
        tot_p = sum(raw_p) or 1.0
        P = [value / tot_p for value in raw_p]
    else:
        P = [1.0 / n] * n
    # Pre-normalize the sparse transition. When there are no usable edges/external inflow,
    # P itself is the fixed point.
    transitions: dict[int, list[tuple[int, float]]] = {}
    for i, edges in out_targets.items():
        total_w = sum(edges.values())
        if total_w > 0:
            transitions[i] = [(j, w / total_w) for j, w in edges.items()]
    active = set(external_inflow)
    active.update(out_targets)
    for edges in out_targets.values():
        active.update(edges)
    if not active:
        return P

    # For isolated nodes with no in/out edges at all, the score is always the same
    # scalar x P[j]. Aggregate hundreds of thousands of isolated nodes into a single
    # inactive_factor state; each round only walks active endpoints, and the full result
    # is materialized only at the end.
    active_p_sum = sum(P[i] for i in active)
    inactive_p_sum = max(0.0, 1.0 - active_p_sum)
    inactive_factor = 1.0
    active_score = {i: P[i] for i in active}

    for _ in range(max_iter):
        dangling_sum = inactive_factor * inactive_p_sum
        for i in active:
            if i not in transitions:
                dangling_sum += active_score[i]

        base_factor = (1.0 - damping) + damping * dangling_sum
        new_active = {i: base_factor * P[i] for i in active}
        for i, edges in transitions.items():
            source_score = active_score[i]
            for j, portion in edges:
                new_active[j] += damping * source_score * portion
        for j, infl in external_inflow.items():
            new_active[j] += damping * infl / n_refs

        delta = sum(abs(new_active[i] - active_score[i]) for i in active)
        delta += abs(base_factor - inactive_factor) * inactive_p_sum
        active_score = new_active
        inactive_factor = base_factor
        if delta < tol:
            break

    return [
        active_score[i] if i in active_score else inactive_factor * P[i]
        for i in range(n)
    ]


def compute_pagerank(symbols: list[dict], refs: list[dict],
                     *, damping: float = 0.85, max_iter: int = 100,
                     tol: float = 1.0e-6,
                     personalization: dict[str, float] | None = None) -> dict[str, float]:
    """Public compatibility layer: run PageRank on the symbol graph, return {symbol_id: score}."""
    scores = _compute_pagerank_scores(
        symbols, refs, damping=damping, max_iter=max_iter, tol=tol,
        personalization=personalization)
    return {_symbol_id(s): scores[i] for i, s in enumerate(symbols)}


def _line_of(sid: str) -> int:
    """Extract the line from a symbol_id (id format path::scope::name::line)."""
    try:
        return int(sid.rsplit("::", 1)[1])
    except (IndexError, ValueError):
        return 1 << 30


def rank_symbols(symbols: list[dict], refs: list[dict], *, top_n: int | None = None,
                 damping: float = 0.85, focus_symbols=None, focus_files=None) -> list[dict]:
    """Rank symbols by importance, return a symbol list with a "rank" score, sorted high to low.

    Each returned dict = the original symbol fields + "rank" (a float score). If top_n is
    given, only the top N are returned.
    focus_symbols/focus_files (queue 4, query-aware): the caller explicitly passes "the
    symbols/files being edited or asked about" to bias ranking toward relevant areas
    (converted into a personalization vector); omitted = the original static
    structural-centrality ranking.
    """
    personalization = _build_personalization(symbols, focus_symbols, focus_files)
    scores = _compute_pagerank_scores(
        symbols, refs, damping=damping, personalization=personalization)
    if top_n is not None:
        # A map usually only needs the top 100~200; don't copy all 570K dicts and sort
        # everything first.
        chosen = nlargest(
            top_n, enumerate(symbols),
            key=lambda item: (scores[item[0]], -item[0]),
        )
        return [dict(s, rank=scores[i]) for i, s in chosen]
    ranked = [dict(s, rank=scores[i]) for i, s in enumerate(symbols)]
    ranked.sort(key=lambda x: x["rank"], reverse=True)
    return ranked
